### PR TITLE
polylithic(?) initialization of Kokkos

### PR DESCRIPTION
**Pull request by janciesko**
Staged initialization: Updates initialize_internal to reflect three stages.

### CODE REVIEW

File: core/src/impl/Kokkos_Core.cpp [hunk 1]
```diff
@@ -373,6 +372,21 @@ void initialize_internal(const InitArguments& args) {
               << std::endl;
   }
 #endif
+}
+
+void pre_initialize(const InitArguments& args) {
+  if (args.disable_warnings) g_show_warnings = false;
+}
+
+void core_initialize(const InitArguments& args) { initialize_backends(args); }
```
Comment (dalg24) [inline core/src/impl/Kokkos_Core.cpp]: I don't see the point of this function.  Do we anticipate ever having anything else than initializing all the backends inside it?  Also I would only expose `pre` and `post`.

File: core/src/impl/Kokkos_Core.cpp [hunk 2]
```diff
@@ -373,6 +372,21 @@ void initialize_internal(const InitArguments& args) {
               << std::endl;
   }
 #endif
+}
+
+void pre_initialize(const InitArguments& args) {
```
Comment (dalg24) [inline core/src/impl/Kokkos_Core.cpp]: Implementation cannot be in the anonymous namespace.

File: core/src/Kokkos_Core.hpp [hunk 3]
```diff
@@ -122,6 +122,10 @@ void initialize(int& narg, char* arg[]);
 
 void initialize(InitArguments args = InitArguments());
 
+void pre_initialize(InitArguments args = InitArguments());
+
+void post_initialize(InitArguments args = InitArguments());
```
Comment (dalg24) [inline core/src/Kokkos_Core.hpp]: Belongs to `Impl::` (or `Experimental::`) namespace.

Comment (crtrott) [reply core/src/Kokkos_Core.hpp]: Impl

Comment (dalg24) [reply core/src/Kokkos_Core.hpp]: Also no default argument.

Comment (janciesko) [reply core/src/Kokkos_Core.hpp]: Fixed.

**Comment by dalg24** [comment]
Please cleanup history before we merge

**Comment by crtrott** [comment]
I rewrote the history for you Damien ;-)

**Pull request merged**